src/server/frontend_wayland/window_wl_surface_role.h

```diff
@@ -111,13 +115,23 @@ class WindowWlSurfaceRole : public WlSurfaceRole
 
     void commit(WlSurfaceState const& state) override;
 
+    /// Locks and returns weak_shell
+    /// Throws std::logic_error if the shell is null
+    auto shell_checked() -> std::shared_ptr<shell::Shell>;
+
+    /// Locks and returns weak_session
+    /// Throws std::logic_error if the session is null
+    auto session_checked() -> std::shared_ptr<scene::Session>;
```

Comment:
These "logic_errors" are sound like precondition failures, and exceptions should be reserved for post-condition failures. Throwing an exception when noticing a the program state is wrong is problematic as the situation is likely not recoverable.

Should the problem be reported by calling `mir::fatal_error()`?

Reply:
Perhaps the surface role should simply hold a strong pointer to the session?

Reply:
That depends. The prior code doesn't force the session to remain, just expects that the session exists during certain operations.